### PR TITLE
MM-53458: Migrate to the v1 API

### PR DESCRIPTION
**Pull request by agnivade**
We use the official firebase client and move off from the legacy API.

The new API uses a service account file and generates a short lived
oauth2 token that gets refreshed periodically.

https://mattermost.atlassian.net/browse/MM-53458

### CODE REVIEW

**Comment by agnivade** [comment]
@enahum - This needs some good amount of testing. I have tested various scenarios of DM message/thread comment notification, with different types of push notifications - generic, full, id_loaded. Everything seems to work. But I'd want another set of testing to have more confidence.

I will share with you the service account file via lastPass.

**Comment by enahum** [comment]
Share over 1password if you can please, I can always generate one only for testing purposes

**Comment by enahum** [comment]
@agnivade I think this doc should also be updated https://developers.mattermost.com/contribute/more-info/mobile/push-notifications/service/

**Review by enahum — APPROVED** [review]
LGTM, once I get the file I can run some manual tests on my end

File: server/android_notification_server.go [hunk 1]
```diff
@@ -4,16 +4,53 @@
...
-	fcm "github.com/appleboy/go-fcm"
+	firebase "firebase.google.com/go/v4"
+	"firebase.google.com/go/v4/messaging"
 	"github.com/kyokomi/emoji"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+const (
+	apnsAuthError       = "APNS_AUTH_ERROR"
```
[the remark this answers is not in the document]
Comment (agnivade) [reply server/android_notification_server.go:23]: Yep, just to handle the error case for completeness.

**Comment by enahum** [comment]
Tested locally, seems to be working just fine

**Comment by agnivade** [comment]
Nice, I will do some more testing while Ibrahim and Eva do their reviews.

**Review by esarafianou — APPROVED** [review]

File: server/android_notification_server.go [hunk 2]
```diff
@@ -77,54 +149,57 @@ func (me *AndroidNotificationServer) SendNotification(msg *PushNotification) Pus
...
 
-	if me.AndroidPushSettings.AndroidAPIKey != "" {
-		sender, err := fcm.NewClient(me.AndroidPushSettings.AndroidAPIKey)
-		if err != nil {
-			if me.metrics != nil {
-				me.metrics.incrementFailure(PushNotifyAndroid, pushType, "invalid ApiKey")
-			}
-			return NewErrorPushResponse(err.Error())
-		}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (agnivade) [inline server/android_notification_server.go:160]: Note: make this configurable.

**Review by isacikgoz — APPROVED** [review]
LGTM, @agnivade did you mean the `v4` API, I didn't get the `v1` API thing.

**Comment by agnivade** [comment]
Ah no, it's a google thing:https://firebase.google.com/docs/cloud-messaging/migrate-v1. They are calling the current API as the legacy API, and the new API as the v1 API. Mainly because it's versioned under the `/v1` path.

**Comment by isacikgoz** [comment]
Ah thanks 👍

**Pull request merged**